networkx/algorithms/isomorphism/isomorphvf2.py

```diff
@@ -577,7 +578,8 @@ def candidate_pairs_iter(self):
             # P(s) = (N_1 - M_1) x {min (N_2 - M_2)}
 
             # elif not (T1_in or T2_in):   # as suggested by  [2], incorrect
-            else:  # as inferred from [1], correct
+            # else:  # as inferred from [1], correct
+            elif not (T2_out or T2_in):
```

Comment:
I haven't followed the logic in depth here, but changing an `else` to an `elif` jumps out to me as having the potential for missing cases. Is it possible to have cases that don't satisfy either of these conditions, or do they cover everything?

Reply:
Thank you for replying so quickly! 🙏

I understand your concern. The 3rd paragraph of the PR description should serve as a proof sketch of correctness, assuming of course that there isn't a flaw in my reasoning and that VF2 is only used for checking if G2 is a subgraph of (or isomorphic to) G1, or enumerating isomorphisms/monomorphisms of G2 and subgraphs of G1, which seems to be the case.

The high-level idea is that, if there's a node v in `T2_out` and `T1_out` is empty, then the current partial mapping can never be extended to a full isomorphism/monomorphism because no node exists in G1 that can be mapped to v, otherwise it would have been included in `T1_out`, by the definition of `T1_out` and `T2_out` in section 2.1 of [Cordella et al](https://www.researchgate.net/profile/Mario-Vento/publication/3193784_A_SubGraph_Isomorphism_Algorithm_for_Matching_Large_Graphs/links/0c960516872cd61d3f000000/A-SubGraph-Isomorphism-Algorithm-for-Matching-Large-Graphs.pdf). The same applies to `T1_in` and `T2_in`.

If there's anything in the proof sketch that doesn't seem right or is not clear feel free to point it out and I'll be happy to check/elaborate on that. 😄

Reply:
That seems reasonable to me, but I'm not confident in my ability to reason through every case *a priori* :)

One thing that could be done just to be sure is to do something like:

```python
else:
    raise nx.NetworkXAlgorithmError("This combination of T1/T2/in/out/inout should not be possible")
```

... with a better exception message of course :).

Reply:
Actually, that combination is the scenario that I described, i.e. `not (T2_out or T2_in)` will be `False` in the scenario that `T1_out` is empty and `T2_out` is not. In this case, there's no point in iterating further candidate pairs because it's impossible to obtain a full feasible mapping from the current partial mapping.

Reply:
Can you rewrite your third paragraph with a little more info to help us get through the logic? In particular:

> Therefore, v will never be included in the current partial mapping because no node of G1 can be mapped to v, otherwise that node would have been included in T1_out.

Why would a node being in `T2_out` without any nodes in `T1_out` mean it is impossible to obtain a full feasible mapping? Are you saying that `T1_out` will never have a node added to it later?

It sounds like you are updating the `feasibility` check to include a new criteria. Cutting earlier will always make it faster, so I'm not surprised by that result. How do you know it is not missing something.  (keep in mind that we didn't write this code, we are maintaining it, so we may need you to spell out more of your argument.)

Also, we are close to implementing the VF2++ algorithm which has a few improvements over VF2 and might even include this one -- but I can't tell yet. :)

Reply:
> Why would a node being in `T2_out` without any nodes in `T1_out` mean it is impossible to obtain a full feasible mapping? Are you saying that `T1_out` will never have a node added to it later?
> 
> It sounds like you are updating the `feasibility` check to include a new criteria. Cutting earlier will always make it faster, so I'm not surprised by that result. How do you know it is not missing something. (keep in mind that we didn't write this code, we are maintaining it, so we may need you to spell out more of your argument.)

From section 2.1 of [Cordella et al.](https://www.researchgate.net/profile/Mario-Vento/publication/3193784_A_SubGraph_Isomorphism_Algorithm_for_Matching_Large_Graphs/links/0c960516872cd61d3f000000/A-SubGraph-Isomorphism-Algorithm-for-Matching-Large-Graphs.pdf), `T1_out` is defined as the set of all nodes of G1, not in the current partial mapping, that are outgoing neighbors of a node in the current partial mapping. Same goes for `T2_out` and G2. If, at any point, a state is reached where `T2_out = [v1, ..., vN]` and `T1_out = []`, then, by definition, there are no nodes left in G1 that can be mapped to `v1, ..., vN`, otherwise those nodes would have been included in `T1_out`. More nodes can be added to `T1_out` throughout the rest of the execution of the algorithm, but the `v1, ..., vN` nodes will remain unmappable.

Consider the following simple example just to illustrate the idea. Let G1 be the graph `u1 -> u2` and G2 the graph `v1 -> v2 -> v3`, and that the current partial mapping is `{(u1, v1), (u2, v2)}`. In this case, we have `T1_out = []` and `T2_out = [v3]`. Because `T1_out` is empty, this means that, by definition, there exists no node in G1 that can be mapped to `v3`.

In fact, after deep diving into the feasibility rules described in the paper, I realized that the scenario that `T2_out or T2_in` is `True` is actually an invalid state when testing for isomorphism, since a candidate pair that would produce such a state does not satisfy the R_new rule, so the exception proposed by @rossbar does make sense in that scenario. That state can still occur when checking for monomorphism though, because the R_in, R_out and R_new rules are not applied when checking for monomorphism (although I believe that at least R_in and R_out could be applied, but that's out of the scope of this PR).

I apologize for the long explanation. It's a bit hard to discuss this kind of stuff in written form. 😅
 
> Also, we are close to implementing the VF2++ algorithm which has a few improvements over VF2 and might even include this one -- but I can't tell yet. :)

Unfortunately, looking at the [VF2++ paper](http://bolyai.cs.elte.hu/egres/tr/egres-18-03.pdf) (more specifically claim 4.4 of section 4.2), it doesn't seem to include this improvement.